lib/spack/spack/build_environment.py

```diff
@@ -474,6 +470,41 @@ def static_to_shared_library(static_lib, shared_lib=None, **kwargs):
     m.static_to_shared_library = static_to_shared_library
 
 
+def get_parallel_make_jobs(pkg):
+    """Returns the number of jobs to be used to build this package
+
+    Args:
+        pkg: package to be built
+    """
+    # If pkg.parallel is False, the package can't be built in parallel
+    if not pkg.parallel:
+        return 1
+
+    # Get the maximum number of build jobs, as defined in the config file
+    max_build_jobs = spack.config.get('config:max_build_jobs') or \
+        multiprocessing.cpu_count()
+
+    # The default number of build jobs is the minimum between max_build_jobs
+    # and the available number of cores
+    default_build_jobs = min(max_build_jobs, multiprocessing.cpu_count())
+
+    # The requested number of build jobs is either pkg.make_jobs (usually
+    # a cli request) or 'config:build_jobs' or the default
+    build_jobs = pkg.make_jobs or spack.config.get('config:build_jobs') \
+        or default_build_jobs
+
+    # If the number of jobs exceeds max_build_jobs, print a warning and
+    # reduce the parallelism
+    if build_jobs > max_build_jobs:
+        msg = 'the number of parallel build jobs requested for package "{0}"' \
+              ' exceeds the maximum number of build jobs allowed ' \
+              '[requested: {1}, maximum: {2}, used: {2}]'
+        tty.warn(msg.format(pkg.name, build_jobs, max_build_jobs))
+        build_jobs = max_build_jobs
```

Comment:
I have some simplifying suggestions for this.

Consider:
1. I think having both `build_jobs` and `max_build_jobs` is confusing and adds complexity.  We should not have two options.
2. I think we should just *never* run with more processes than `multiprocessing.cpu_count()`.  Can you think of a case where that is a reasonable thing to do?

Given both of those things, how about we just make `multiprocessing.cpu_count()` the hard maximum, unless `build_job` is set to _less_ than that.

That simplifies the logic in this function to:

```
build_jobs min(spack.config.get('config:build_jobs'), multiprocessing.cpu_count())
```

And it's much easier to explain -- there's no "request"; we let the config system handle that as it should.

I think we can get rid of the warning above.  I think it's just extra noise and it'll get old for users.  Could we convert it to a debug message?

Reply:
Fine with me. I assume this also means we'll change the default in `config.yaml` to be:
```yaml
config:
  build_jobs: 16
```
in order to fix #11072, is that right?

Reply:
@alalazo: yes

@healther: If we really want to make the user able to override, we *could* add something like `config.get_blame_scope(‘config:build_jobs’)` to check if it was set on the CLI, and allow manual overrides.

I think it’s a minor use case though, especially since `multiprocessing.cpu_count()` usually reports number of hyperthreads (not physical cores) anyway.

Reply:
Agreed, I was only mentioning it, because currently we are trying to push down our build time (to get some reasonable turnaround times for our software environment deployment). We are currently far away from wanting to optimise build times of single packages on the level of oversubscribing cores, but at some point someone might want to chase down these last percentages.

For the moment I'd be more than happy with a warning if a user wants more than the system provides and maybe a comment on how to get around it (like a pointer to the code in ?config? where the parameter handling will be done).

Reply:
Implemented in 1c5fb9b135559ba0a584bb4ea4c071ba063d2d59